transformers/modeling_unilm.py

```diff
@@ -0,0 +1,747 @@
+        self.forbid_ignore_set = forbid_ignore_set
+        self.ngram_size = ngram_size
+        self.min_len = min_len
+        self.init_weights()
+        self.tie_weights()
+
+    def tie_weights(self):
+        """ Make sure we are sharing the input and output embeddings.
+            Export to TorchScript can't handle parameter sharing so we are cloning them instead.
+        """
+        self._tie_or_clone_weights(self.cls.predictions.decoder,
+                                   self.bert.embeddings.word_embeddings)
+
+    def forward(self, input_ids, token_type_ids, position_ids, attention_mask):
+        if self.search_beam_size > 1:
+            return self.beam_search(input_ids, token_type_ids, position_ids, attention_mask)
+
+        input_shape = list(input_ids.size())
+        batch_size = input_shape[0]
+        input_length = input_shape[1]
+        output_shape = list(token_type_ids.size())
+        output_length = output_shape[1]
+
+        output_ids = []
+        prev_embedding = None
+        prev_encoded_layers = None
+        curr_ids = input_ids
+        mask_ids = input_ids.new(batch_size, 1).fill_(self.mask_word_id)
+        next_pos = input_length
+
+        while next_pos < output_length:
+            curr_length = list(curr_ids.size())[1]
+
+            start_pos = next_pos - curr_length
+            x_input_ids = torch.cat((curr_ids, mask_ids), dim=1)
+
+            curr_token_type_ids = token_type_ids[:, start_pos:next_pos+1]
+            curr_attention_mask = attention_mask[:,
+                                                 start_pos:next_pos+1, :next_pos+1]
+            curr_position_ids = position_ids[:, start_pos:next_pos+1]
+            new_embedding, new_encoded_layers, _ = \
+                self.bert(x_input_ids, curr_token_type_ids, curr_position_ids, curr_attention_mask,
+                          output_all_encoded_layers=True, prev_embedding=prev_embedding, prev_encoded_layers=prev_encoded_layers)
+
+            last_hidden = new_encoded_layers[-1][:, -1:, :]
+            prediction_scores = self.cls(last_hidden)
+            _, max_ids = torch.max(prediction_scores, dim=-1)
+            output_ids.append(max_ids)
+
+            if prev_embedding is None:
+                prev_embedding = new_embedding[:, :-1, :]
+            else:
+                prev_embedding = torch.cat(
+                    (prev_embedding, new_embedding[:, :-1, :]), dim=1)
+            if prev_encoded_layers is None:
+                prev_encoded_layers = [x[:, :-1, :]
+                                       for x in new_encoded_layers]
+            else:
+                prev_encoded_layers = [torch.cat((x[0], x[1][:, :-1, :]), dim=1)
+                                       for x in zip(prev_encoded_layers, new_encoded_layers)]
+            curr_ids = max_ids
+            next_pos += 1
+
+        return torch.cat(output_ids, dim=1)
+
+    def beam_search(self, input_ids, token_type_ids, position_ids, attention_mask):
```

Comment:
possible to use `PretrainedModel.generate`?

https://github.com/huggingface/transformers/blob/6c1b23554f8bb5b5e1f6c80969acab764c755678/src/transformers/modeling_utils.py#L585

Reply:
either for this or forward